### ISSUE TITLE
Run/Restart/Debug game right from Rider

### ISSUE THREAD

**Issue by ArtCraft** [JetBrains/Unity3dRider#34]
It would be nice to have a way to Build/Run/Restart/Debug unity game right from Rider 

maybe it's my fault, but it is unclear for me how to configure mono remote

**Comment by citizenmatt** [comment]
I think this would need to use Unity's REST API, which is mentioned in the [roadmap](https://unity3d.com/unity/roadmap) (and I think might be used in Visual Studio Tools for Unity - they definitely "connect" to Unity) but I don't see it documented anywhere.

**Comment by citizenmatt** [comment]
Although if you want to debug an existing, running instance of Unity, just use the _Attach to Local Process_ menu item, and it will list Unity processes.

**Comment by ArtCraft** [comment]
> Although if you want to debug an existing, running instance of Unity, just use the Attach to Local Process menu item, and it will list Unity processes.

I've found a comand to get list of unity processes here: http://stackoverflow.com/questions/35768547/unity-debug-with-rider-eap-on-windows

`netstat -ano | ?{($_ -like "*"+$(Get-Process -Name Unity).Id) -and ($_ -like "* LISTENING *") }`

But as i understand, port may change from time to time.
So I have to dig for port every time...

Is it possible to configure it automatically?

**Comment by van800** [comment]
In Rider menu just try:
Run → "Attach to local process" menu item to list available Unity instances.

**Comment by ArtCraft** [comment]
Wate a moment, why this is closed? 
Ok, there is a way to debug already running game...
But there is no way to run/build/restart game from rider...

**Comment by ArtCraft** [comment]
in unity there is a command for it by the way
`EditorApplication.ExecuteMenuItem("Edit/Play");`

https://docs.unity3d.com/ScriptReference/EditorApplication.ExecuteMenuItem.html

**Comment by ArtCraft** [comment]
шахов, какого черта, я ведь знаю где тебя найти )

**Comment by van800** [comment]
related to https://youtrack.jetbrains.com/issue/RIDER-3260
https://github.com/JetBrains/Unity3dRider/issues/22

**Comment by van800** [comment]
@ArtCraft How can 
EditorApplication.ExecuteMenuItem("Edit/Play"); 
be executed from other process like Rider. Please share, if you know. Thanks!

**Comment by ArtCraft** [comment]
you already have an editor script that users have to place in their project, so the only thing needed to make it possible for Rider to call any Unity editor commands,  is some sort of inter-process communication like socket

**Comment by van800** [comment]
Thanks!
I see that VS has feature "Debug Unity and Play" kind of similar to the one proposed.

**Comment by van800** [comment]
I have done some experiments in branch https://github.com/JetBrains/Unity3dRider/tree/udp.run with passing commands via socket, however can't find an easy way to dispatch the call to main thread.
The drawback of solutions like http://www.what-could-possibly-go-wrong.com/the-dispatcher-pattern/ requires attaching MonoBehaviour to some GameObject, which I don't want to do.
Am I missing something?

**Comment by ArtCraft** [comment]
maybe this can be helpfull 
https://github.com/PimDeWitte/UnityMainThreadDispatcher

**Comment by van800** [comment]
This feature will appear in Rider 2018.1 EAP1 in a week or two.

**Comment by NIA** [comment]
@van800 The latest blog post mentions that this feature is indeed already included in 2018.1 EAP1. But where can these buttons be found? When I open the project I cannot see it anywhere in menus or debug tool window.

I did removed the `Unity3dRider.cs` as suggested in a blog post and placed a `JetBrains.Rider.Unity.Editor.Plugin.Repacked.dll` there, the new plugin is functional (e.g., the new Unity Explorer is working)

**Comment by van800** [comment]
![image](https://user-images.githubusercontent.com/1482681/36787209-96138d2a-1c89-11e8-9cda-c6cc80cfed47.png)
Please check buttons in the left menu.

**Comment by NIA** [comment]
@van800 hmm, I do not seem to see "Unity" tool window...
![image](https://user-images.githubusercontent.com/26939/36787397-156ec100-1c9b-11e8-94b1-031cbfd3eb12.png)

Am I missing something? 😕 

Also "Attach to Unity Editor & Play" run configuration, when started, definitely attaches, but not runs.

**Comment by van800** [comment]
The toolwindow appears only after connection between Unity and Rider is established.

**Comment by NIA** [comment]
@van800 So I have problems with plugin, right? I can "Attach to Unity editor" and successfully debug the application, so the connection somehow exists, but still no tool window.

Also strange: each time I start Rider (or just reopen the solution), the same message is shown each time in Event Log and popup:

> 	Unity: plugin installed
> 			Rider plugin v2.1.3.2067 for the Unity Editor was automatically installed for the project 'dto_unity'
> 			This allows better integration between the Unity Editor and Rider IDE.
> 			The plugin file can be found on the following path:
> 			..\Assets\Plugins\Editor\JetBrains\Unity3DRider.cs.
> 			Please switch back to Unity to make plugin file appear in the solution.

...although it was already placed into Assets\Plugins\Editor\JetBrains on first solution opening. Somehow Rider did not remember that it already installed the plugin?

_(also note the old file name in the message – Unity3DRider.cs – just cosmetics, I guess, but still)_

**Comment by van800** [comment]
It doesn't seem right, 2018.1 EAP1 should already install the dll, not cs. I will check.

**Comment by NIA** [comment]
@val800 it actually installs dll. Just the message is old

**Comment by van800** [comment]
![image](https://user-images.githubusercontent.com/1482681/36788324-c6325ad2-1c8d-11e8-88aa-9e2befdfe7aa.png)
To check the connection you would need to set logging level to something like Verbose.
After that try to double click any script in Unity, find message like 
`Rider plugin initialized. LoggingLevel: TRACE. Change it in Unity Preferences -> Rider. Logs path: /var/folders/by/f499_4kj7075g5j07b82qp_c0000gn/T/Unity3dRider/2018-02-28T-13-33-08.log.
`
and upload here the log file.
Thanks!

**Comment by van800** [comment]
Yeah, the issue about showing "plugin installed" every time is a known issue. Already fixed for EAP2.

**Comment by NIA** [comment]
Rider plugin initialized. LoggingLevel: VERBOSE. Change it in Unity Preferences -> Rider. Logs path: C:\Users\Ivan\AppData\Local\Temp\Unity3dRider\2018-02-28T-15-53-59.log.
    UnityEngine.Debug:Log(Object)
    JetBrains.Rider.Unity.Editor.PluginEntryPoint:Init()
    JetBrains.Rider.Unity.Editor.PluginEntryPoint:.cctor()
    UnityEditor.EditorAssemblies:ProcessInitializeOnLoadAttributes()

The file is attached: [2018-02-28T-15-53-59.log](https://github.com/JetBrains/Unity3dRider/files/1767117/2018-02-28T-15-53-59.log)

**Comment by van800** [comment]
Ah, I am sorry, could you please switch it to Trace and send again. And add the Rider logs from Help->Show log in Explorer. Zip all logs files in the folder!

**Comment by NIA** [comment]
Tried another project (ExampleScene, that was on screenshot). Same picture. No Unity tab, then set TRACE level close unity, open unity, close rider, open rider ⇒ Unity tab appears. Logs appear, buttons work:
![image](https://user-images.githubusercontent.com/26939/36789752-aca80434-1ca3-11e8-87f2-74d1fd56e629.png)
How can this even be possible?

**Comment by van800** [comment]
Hmm ) I have always trace logs enabled. I will investigate. Thanks!

**Comment by van800** [comment]
if it is still an issue - please create a new one in https://github.com/JetBrains/resharper-unity
